File: backend/app/routers/experience.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.experience_db import ExperienceDatabaseError
from app.services import experience_db_store, experience_service, settings_service, vector_search
from app.services.experience_service import ExperienceExtractionError

router = APIRouter(prefix="/api/experience", tags=["experience"])


class SaveDatabaseRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # Raw editor text, so JSON syntax errors can be reported with line/column.
    text: str


class ExtractRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    jobId: str
    jobDescription: str
    jobTitle: str = ""
    jobMission: str = ""
    techSkills: list[str] = Field(default_factory=list)


def _bad_request(code: str, message: str) -> HTTPException:
    return HTTPException(status_code=400, detail={"code": code, "message": message})


def _resolve_profile(profile_id: str | None) -> str:
    """Which profile's corpus to act on. Defaults to the active one."""
    from app.services import job_store

    if profile_id:
        return profile_id
    try:
        return str(job_store.active_profile_id())
    except job_store.NoProfile as exc:
        raise _bad_request("NO_PROFILE", str(exc)) from exc


@router.get("/database")
def get_database(profileId: str | None = None):
    """One profile's corpus: raw text for the editor, plus its company list."""
    target = _resolve_profile(profileId)
    text = experience_db_store.load_raw(target)

    if not text:
        # No corpus yet is a normal state for a new profile, not an error.
        return {
            "text": "",
            "companies": [],
            "path": str(experience_db_store.path(target)),
            "profileId": target,
            "exists": False,
            "valid": True,
            "detail": None,
        }

    try:
        db = experience_db_store.load_database(target)
    except ExperienceDatabaseError as exc:
        # Still return the text so the user can fix it in the editor.
        return {
            "text": text,
            "companies": [],
            "path": str(experience_db_store.path(target)),
            "profileId": target,
            "exists": True,
            "valid": False,
            "detail": str(exc),
        }

    return {
        "text": text,
        "companies": db.company_names(),
        "path": str(experience_db_store.path(target)),
        "profileId": target,
        "exists": True,
        "valid": True,
        "detail": None,
    }


@router.put("/database")
def save_database(payload: SaveDatabaseRequest, profileId: str | None = None):
    target = _resolve_profile(profileId)
    try:
        db = experience_db_store.save_raw_text(target, payload.text)
    except ExperienceDatabaseError as exc:
        raise _bad_request("INVALID_DATABASE", str(exc)) from exc
    # Company list is returned so the dropdown updates without a second call.
    return {
        "companies": db.company_names(),
        "profileId": target,
        "exists": True,
        "valid": True,
        "detail": None,
    }


@router.get("/database/example")
def database_example():
    """The expected shape, for a profile starting from nothing."""
    import json

    return {"text": json.dumps(experience_db_store.SEED, indent=2, ensure_ascii=False)}


@router.get("/progress")
def get_progress(since: int = 0):
    """Events newer than `since`, for the live console."""
    from app.services.progress import progress

    return {"events": progress.since(since), "latest": progress.latest_seq()}


@router.post("/progress/clear")
def clear_progress():
    from app.services.progress import progress

    progress.clear()
    return {"ok": True, "latest": progress.latest_seq()}


@router.get("/search-backend")
def search_backend():
    """Whether ranking is semantic or the lexical fallback."""
    return vector_search.backend()


@router.get("/all")
def all_extractions():
    """Every stored extraction, so the table can restore badges after reload."""
    return experience_service.all_experience()


@router.get("/{job_id}")
def get_extraction(job_id: str):
    found = experience_service.get_experience(job_id)
    if found is None:
        raise HTTPException(
            status_code=404,
            detail={"code": "NOT_EXTRACTED", "message": "No experience extracted for this job."},
        )
    return found


@router.post("/extract")
async def extract(payload: ExtractRequest):
    # An applied job is a record of what was sent. Re-extracting would rewrite
    # the bullets behind a submitted application, so the lock is enforced here
    # rather than relying on the button being disabled.
    from app.services import job_store

    if job_store.is_locked(payload.jobId):
        raise _bad_request(
            "JOB_LOCKED",
            "This job is marked applied, so its experience can no longer be changed.",
        )

    first_company = (settings_service.get_settings().get("firstCompany") or "").strip()
    if not first_company:
        raise _bad_request(
            "NO_FIRST_COMPANY", "Please select a First Company in Settings first."
        )

    target = _resolve_profile(None)
    try:
        db = experience_db_store.load_database(target)
    except experience_db_store.CorpusNotFound as exc:
        raise _bad_request(
            "NO_CORPUS",
            "This profile has no database.json yet. Add one on the Profile tab "
            "before extracting.",
        ) from exc
    except ExperienceDatabaseError as exc:
        raise _bad_request("INVALID_DATABASE", str(exc)) from exc

    # The title step is asked to improve on what the profile already claims, so
    # it needs to know what that is.
    from app.services import profile_service

    try:
        current_title = profile_service.get_profile(target).data.profile.professionalTitle
    except profile_service.ProfileNotFound:
        current_title = ""

    try:
        result = await experience_service.extract_experience(
            db=db,
            first_company=first_company,
            job_description=payload.jobDescription,
            tech_skills=payload.techSkills,
            job_title=payload.jobTitle,
            job_mission=payload.jobMission,
            current_title=current_title,
        )
    except ExperienceExtractionError as exc:
        raise _bad_request("EXTRACTION_FAILED", str(exc)) from exc

    experience_service.save_experience(payload.jobId, result)
    return result

File: frontend/src/api/experience.ts
import axios from "axios";
import { BACKEND_URL } from "../config";

export interface ExperienceJob {
  company: string;
  product: string;
  timeline: string;
  companySummary: string;
  projects: string[];
  bullets: string[];
  source_challenge_ids: string[];
}

export interface ExperienceResult {
  job1: ExperienceJob;
  job2: ExperienceJob;
  /** Resume summary written from the bullets above; "" when unavailable. */
  summary: string;
  summarySource: "deepseek" | "none";
  /** Resume headline finalized by ChatGPT from a DeepSeek draft; "" when unavailable. */
  title: string;
  titleSource: "chatgpt" | "none";
  /** Resume skill set written for this job, last in the DeepSeek chat;
   *  [] when unavailable (the profile's own skills are used instead). */
  skillSet: string[];
  skillSetSource: "deepseek" | "none";
  /** ChatGPT's categorization of skillSet, from the same revision pass that
   *  handles bullets/summaries; [] when it never ran or didn't parse, in
   *  which case skillSet renders as one uncategorized group instead. */
  skillGroups: { category: string; skills: string[] }[];
  /** Prompts that shared this job's single DeepSeek chat; 0 = never connected. */
  deepseekTurns: number;
  search: { mode: "semantic" | "lexical"; model: string | null; detail: string | null };
  /** "fallback" means the AI provider was unavailable and bullets were composed
   *  directly from database.json rather than generated. "chatgpt" means
   *  DeepSeek generated the bullets and summary, and a final ChatGPT pass
   *  then revised them — that revised text is what's actually on the resume. */
  generator: "deepseek" | "chatgpt" | "fallback";
  extractedAt: string;
}

/** One profile's career corpus. Each profile has its own database.json. */
export interface DatabaseInfo {
  text: string;
  companies: string[];
  path: string;
  profileId: string;
  /** False when this profile has no corpus yet — a normal state, not an error. */
  exists: boolean;
  valid: boolean;
  detail: string | null;
}

export async function fetchExperienceDatabase(profileId?: string): Promise<DatabaseInfo> {
  const response = await axios.get<DatabaseInfo>(`${BACKEND_URL}/api/experience/database`, {
    params: profileId ? { profileId } : undefined,
  });
  return response.data;
}

export async function saveExperienceDatabase(
  text: string,
  profileId?: string,
): Promise<{ companies: string[]; profileId: string }> {
  const response = await axios.put<{ companies: string[]; profileId: string }>(
    `${BACKEND_URL}/api/experience/database`,
    { text },
    { params: profileId ? { profileId } : undefined },
  );
  return response.data;
}

/** The expected shape, for a profile starting from nothing. */
export async function fetchDatabaseExample(): Promise<string> {
  const response = await axios.get<{ text: string }>(
    `${BACKEND_URL}/api/experience/database/example`,
  );
  return response.data.text;
}

export async function extractExperience(payload: {
  jobId: string;
  jobDescription: string;
  jobTitle?: string;
  jobMission?: string;
  techSkills?: string[];
}): Promise<ExperienceResult> {
  const response = await axios.post<ExperienceResult>(
    `${BACKEND_URL}/api/experience/extract`,
    payload,
  );
  return response.data;
}

/** Every stored extraction, so badges survive a page refresh. */
export async function fetchAllExperience(): Promise<Record<string, ExperienceResult>> {
  const response = await axios.get<Record<string, ExperienceResult>>(
    `${BACKEND_URL}/api/experience/all`,
  );
  return response.data;
}
